include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,182 @@
+	/** External supply is present and of programmable output*/
+	CHARGER_ONLINE_PROGRAMMABLE,
+};
+
+/**
+ * @brief Charging states
+ */
+enum charger_status {
+	/** Charging device state is unknown */
+	CHARGER_STATUS_UNKNOWN = 0,
+	/** Charging device is charging a battery */
+	CHARGER_STATUS_CHARGING,
+	/** Charging device is not able to charge a battery */
+	CHARGER_STATUS_DISCHARGING,
+	/** Charging device is not charging a battery */
+	CHARGER_STATUS_NOT_CHARGING,
+	/** The battery is full and the charging device will not attempt charging */
+	CHARGER_STATUS_FULL,
+};
+
+/**
+ * @typedef charger_property_t
+ * @brief A charger property's identifier
+ *
+ * See charger_props for a list of identifiers
+ */
+typedef uint16_t charger_property_t;
+
+/**
+ * @brief Representation of a charger property value
+ */
+union charger_value {
+	enum charger_online online;
+	bool present;
+	enum charger_status status;
+	int custom_val;
+};
+
+/**
+ * @typedef charger_value_t
+ * @brief A type representing charger_value
+ */
+typedef union charger_value charger_value_t;
+
+/**
+ * @typedef charger_get_property_t
+ * @brief Callback API for getting a charger property.
+ *
+ * See charger_get_property() for argument description
+ */
+typedef int (*charger_get_property_t)(const struct device *dev, charger_property_t prop,
+				      charger_value_t *val);
+
+/**
+ * @brief Charging device API
+ *
+ * Caching is entirely on the onus of the client
+ */
+__subsystem struct charger_driver_api {
+	charger_get_property_t get_property;
+};
+
+/**
+ * @brief Fetch a battery charger property
+ *
+ * @param dev Pointer to the battery charger device
+ * @param prop pointer to charger_get_property struct where the property struct
+ * field is set by the caller to determine what property is read from the
+ * charger device into the charger_get_property struct's value field.
+ *
+ * @retval 0 if successful
+ * @retval < 0 if getting property failed
+ */
+__syscall int charger_get_prop(const struct device *dev, charger_property_t prop,
+			       charger_value_t *val);
+
+static inline int z_impl_charger_get_prop(const struct device *dev, charger_property_t prop,
+					  charger_value_t *val)
+{
+	const struct charger_driver_api *api = (const struct charger_driver_api *)dev->api;
+
+	return api->get_property(dev, prop, val);
+}
+
+/**
+ * @brief Fetch multiple battery charger properties
+ *
+ * @param dev Pointer to the battery charger device
+ * @param props pointer to array of charger_get_property struct where the property struct
+ * field is set by the caller to determine what property is read from the
+ * charger device into the charger_get_property struct's value field. The array is not
+ * permuted and elements can be directly accessed as loaded.
+ * @param props_len number of properties in props array
+ *
+ * @retval 0 if successful
+ * @retval < 0 if getting a property failed
+ */
+__syscall int charger_get_props(const struct device *dev, charger_property_t *props,
```

Comment:
I think I missed some conversation here.

I really love the idea of supporting multi-get properties. But if we do that we need a way to find the return codes for each property. Or at very least to know why some property is failing and what that property actually is.

Right now we *cannot* find out which properties failed and why when using `charger_get_props()`.

I see two options:

1. Add another parallel array that contains return codes.
2. Go back to using a struct that embeds this information with it's value.
3. Let's leave out multi-get for future work on this API. We don't actually need it for an MVP.

I prefer 3) so we can take our time discussing after code-freeze and we have something minimal merged.

*As indicated by the fuel-gauge API I'm a fan of option 2)* because parallel arrays requires more memory management by the client instead of the compiler where we're more likely to run into potential memory issues. We can still do 1), but we need to do one or the other otherwise the API as it stands isn't usable.

We also have the bonus option 3) 

@henrikbrixandersen @gmarull @teburd thoughts here? I know @rriveramcrus is AFK atm and figure I should start some discussion on it.

Reply:
> I prefer 3) so we can take our time discussing after code-freeze and we have something minimal merged.

I see your points here. If the MVP does not require multi-prop gets, I think we should go with 3) for now.

Reply:
Agree with @henrikbrixandersen , drop it and come back to it

Reply:
Went ahead and pushed option 3. Thanks @aaronemassey for putting the possible options on the table.